Flaky-test: KafkaMessageOrderKafkaTest.testKafkaProduceMessageOrder

KafkaMessageOrderKafkaTest is flaky. The testKafkaProduceMessageOrder test method fails sporadically.

[example failure](https://github.com/streamnative/kop/runs/7738563391?check_suite_focus=true#step:12:207)

```
Error:  testKafkaProduceMessageOrder(io.streamnative.pulsar.handlers.kop.KafkaMessageOrderKafkaTest)  Time elapsed: 0.194 s  <<< FAILURE!
java.lang.AssertionError: expected [0] but found [19]
	at org.testng.Assert.fail(Assert.java:96)
	at org.testng.Assert.failNotEquals(Assert.java:776)
	at org.testng.Assert.assertEqualsImpl(Assert.java:137)
	at org.testng.Assert.assertEquals(Assert.java:118)
	at org.testng.Assert.assertEquals(Assert.java:652)
	at org.testng.Assert.assertEquals(Assert.java:662)
	at io.streamnative.pulsar.handlers.kop.KafkaMessageOrderTestBase.testKafkaProduceMessageOrder(KafkaMessageOrderTestBase.java:170)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
@@ -1125,23 +1125,15 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
                     partitionData.complete(Pair.of(Errors.UNKNOWN_SERVER_ERROR, -1L));
                     return;
                 }
-
                 if (log.isDebugEnabled()) {
                     log.debug("[{}] Get earliest position for topic {}: {}, lac: {}",
                             ctx, perTopic.getName(), position, lac);
                 }
-                final long latestOffset = MessageMetadataUtils.getCurrentOffset(managedLedger);
-                if (latestOffset < 0) {
-                    log.warn("[{}] Unexpected latest offset {} (< 0) for topic {}",
-                            ctx, latestOffset, perTopic.getName());
-                    partitionData.complete(Pair.of(Errors.NONE, 0L));
-                    return;
-                }
                 if (position.compareTo(lac) > 0) {
-                    partitionData.complete(Pair.of(Errors.NONE, latestOffset));
+                    partitionData.complete(Pair.of(Errors.NONE, 0L));
                 } else {
                     MessageMetadataUtils.getOffsetOfPosition(managedLedger, position, false,
-                            timestamp, skipMessagesWithoutIndex)
+                                    timestamp, skipMessagesWithoutIndex)
                             .whenComplete((offset, throwable) -> {
                                 if (throwable != null) {
                                     log.error("[{}] Failed to get offset for position {}",
@@ -1149,10 +1141,13 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
                                     partitionData.complete(Pair.of(Errors.UNKNOWN_SERVER_ERROR, null));
                                     return;
                                 }
+                                if (log.isDebugEnabled()) {
+                                    log.debug("[{}] Get offset of position for topic {}: {}, lac: {}, offset: {}",
+                                            ctx, perTopic.getName(), position, lac, offset);
+                                }
                                 partitionData.complete(Pair.of(Errors.NONE, offset));
-                    });
+                            });
                 }
-
             } else {
                 fetchOffsetByTimestamp(partitionData, managedLedger, lac, timestamp, perTopic.getName());
             }

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaMessageOrderTestBase.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaMessageOrderTestBase.java
@@ -164,7 +164,10 @@ public abstract class KafkaMessageOrderTestBase extends KopProtocolHandlerTestBa
             kConsumer.getConsumer().subscribe(Collections.singleton(topicName));
             int[] receivedKeys = new int[totalMsgs];
             for (int i = 0; i < totalMsgs; ) {
-                ConsumerRecords<Integer, String> records = kConsumer.getConsumer().poll(Duration.ofSeconds(1));
+                ConsumerRecords<Integer, String> records = kConsumer.getConsumer().poll(Duration.ofSeconds(5));
+                if (records.isEmpty()) {
+                    break;
+                }
                 for (ConsumerRecord<Integer, String> record : records) {
                     if (log.isDebugEnabled()) {
                         log.debug("Kafka consumer get i: {} offset: {}, message: {}, key: {}",

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaMessageOrderTestBase.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaMessageOrderTestBase.java
@@ -162,17 +162,20 @@ public abstract class KafkaMessageOrderTestBase extends KopProtocolHandlerTestBa
             @Cleanup
             KConsumer kConsumer = new KConsumer(topicName, getKafkaBrokerPort(), "testKafkaProduce-KafkaConsume");
             kConsumer.getConsumer().subscribe(Collections.singleton(topicName));
+            int[] receivedKeys = new int[totalMsgs];
             for (int i = 0; i < totalMsgs; ) {
                 ConsumerRecords<Integer, String> records = kConsumer.getConsumer().poll(Duration.ofSeconds(1));
-                if (log.isDebugEnabled()) {
-                    for (ConsumerRecord<Integer, String> record : records) {
-                        log.debug("Kafka consumer get i: {} message: {}, key: {}", i, record.value(), record.key());
-                        assertEquals(record.key().intValue(), i);
-                        i++;
+                for (ConsumerRecord<Integer, String> record : records) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Kafka consumer get i: {} offset: {}, message: {}, key: {}",
+                                i, record.offset(), record.value(), record.key());
                     }
-                } else {
-                    i += records.count();
+                    receivedKeys[i++] = record.key();
                 }
+            }
+            log.info("Received keys: {}", receivedKeys);
+            for (int i = 0; i < totalMsgs; i++) {
+                assertEquals(receivedKeys[i], i);
             }
         } finally {
             if (consumer != null) {

